File: firebase/functions/src/common/money.ts
/**
 * Money helpers. All balances are stored as integer **paise** to avoid
 * floating-point drift. ₹1 = 100 paise. ₹9/min = 900 paise/min = 15 paise/sec.
 *
 * These are pure functions — no Firebase — so they are unit-tested directly.
 */

export const PAISE_PER_RUPEE = 100;
export const SECONDS_PER_MINUTE = 60;

/** Convert rupees (may be fractional) to whole paise. */
export function rupeesToPaise(rupees: number): number {
  return Math.round(rupees * PAISE_PER_RUPEE);
}

/** Convert paise to rupees (for display / logging only). */
export function paiseToRupees(paise: number): number {
  return paise / PAISE_PER_RUPEE;
}

/**
 * Price per second in paise from a per-minute price in paise.
 * We do NOT round here; billing multiplies by whole seconds and rounds once,
 * which keeps long sessions exact (900/60 = 15 exactly for the default price,
 * and fractional prices stay accurate over the whole session).
 */
export function pricePerSecond(pricePerMinutePaise: number): number {
  return pricePerMinutePaise / SECONDS_PER_MINUTE;
}

/**
 * Charge (paise, integer) for a whole number of billed seconds. Rounds the
 * final amount once so we never accumulate sub-paise error.
 */
export function chargeForSeconds(pricePerMinutePaise: number, seconds: number): number {
  if (seconds <= 0) return 0;
  return Math.round(pricePerSecond(pricePerMinutePaise) * seconds);
}

/**
 * Seconds of talk-time affordable with a given spendable balance at a given
 * per-minute price. Floors — you only get whole seconds you can fully pay for.
 */
export function affordableSeconds(spendablePaise: number, pricePerMinutePaise: number): number {
  const pps = pricePerSecond(pricePerMinutePaise);
  if (pps <= 0) return 0;
  return Math.floor(spendablePaise / pps);
}

/** Clamp a value to be non-negative (wallets can never go below zero). */
export function clampNonNegative(value: number): number {
  return value < 0 ? 0 : value;
}

File: packages/shared_flutter/lib/src/utils/money.dart
import 'package:intl/intl.dart';

/// Money helpers mirroring the backend. All balances travel as integer **paise**
/// so the client and Cloud Functions format money identically and never drift.
abstract final class Money {
  static const int paisePerRupee = 100;
  static const int secondsPerMinute = 60;

  static final NumberFormat _inr = NumberFormat.currency(
    locale: 'en_IN',
    symbol: '₹',
    decimalDigits: 0,
  );

  /// Whole-rupee display for balances/plans, e.g. 11900 → "₹119".
  static String formatPaise(int paise) => _inr.format(paise / paisePerRupee);

  /// Two-decimal display where precision matters (receipts), e.g. "₹119.00".
  static String formatPaisePrecise(int paise) {
    final f = NumberFormat.currency(locale: 'en_IN', symbol: '₹', decimalDigits: 2);
    return f.format(paise / paisePerRupee);
  }

  static double paiseToRupees(int paise) => paise / paisePerRupee;
  static int rupeesToPaise(num rupees) => (rupees * paisePerRupee).round();

  /// Seconds of talk-time a spendable balance affords at a per-minute price.
  static int affordableSeconds(int spendablePaise, int pricePerMinutePaise) {
    if (pricePerMinutePaise <= 0) return 0;
    final pps = pricePerMinutePaise / secondsPerMinute;
    return (spendablePaise / pps).floor();
  }

  /// mm:ss formatting for the consultation timer.
  static String formatDuration(int seconds) {
    final s = seconds < 0 ? 0 : seconds;
    final m = (s ~/ 60).toString().padLeft(2, '0');
    final sec = (s % 60).toString().padLeft(2, '0');
    return '$m:$sec';
  }

  /// Human duration for history rows, e.g. "12m 04s" or "1h 03m".
  static String formatDurationLong(int seconds) {
    final h = seconds ~/ 3600;
    final m = (seconds % 3600) ~/ 60;
    final s = seconds % 60;
    if (h > 0) return '${h}h ${m.toString().padLeft(2, '0')}m';
    return '${m}m ${s.toString().padLeft(2, '0')}s';
  }
}
